File: practice-projects/springboot4ddd/src/main/java/com/github/microwind/springboot4ddd/domain/model/order/Order.java
package com.github.microwind.springboot4ddd.domain.model.order;

import com.github.microwind.springboot4ddd.domain.event.DomainEvent;
import com.github.microwind.springboot4ddd.domain.event.order.OrderCancelledEvent;
import com.github.microwind.springboot4ddd.domain.event.order.OrderCompletedEvent;
import com.github.microwind.springboot4ddd.domain.event.order.OrderCreatedEvent;
import com.github.microwind.springboot4ddd.domain.event.order.OrderPaidEvent;
import lombok.Getter;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * 订单领域模型（聚合根）
 *
 * <p>纯领域模型，零持久化注解、零框架依赖。状态字段直接持有 {@link OrderStatus}，
 * 只能通过 {@link #pay()} / {@link #cancel()} / {@link #complete()} 等行为方法
 * 迁移，不暴露任何 setter。
 *
 * <p>持久化重建走 {@link #restore} 静态工厂；数据库自增主键回填与创建事件补登
 * 合并为 {@link #markCreated(Long)}，仅供 infrastructure 层在 save 完成后调用一次。
 *
 * @author jarry
 * @since 1.0.0
 */
@Getter
public class Order {

    private Long id;
    private final String orderNo;
    private final Long userId;
    private final BigDecimal totalAmount;
    private OrderStatus status;
    private final LocalDateTime createdAt;
    private LocalDateTime updatedAt;

    private final List<DomainEvent> domainEvents = new ArrayList<>();

    public enum OrderStatus {
        PENDING("待支付"),
        PAID("已支付"),
        CANCELLED("已取消"),
        COMPLETED("已完成");

        private final String description;

        OrderStatus(String description) {
            this.description = description;
        }

        public String getDescription() {
            return description;
        }
    }

    Order(Long id,
          String orderNo,
          Long userId,
          BigDecimal totalAmount,
          OrderStatus status,
          LocalDateTime createdAt,
          LocalDateTime updatedAt) {
        this.id = id;
        this.orderNo = orderNo;
        this.userId = userId;
        this.totalAmount = totalAmount;
        this.status = status;
        this.createdAt = createdAt;
        this.updatedAt = updatedAt;
    }

    public static Order create(Long userId, BigDecimal totalAmount) {
        LocalDateTime now = LocalDateTime.now();
        return new Order(
                null,
                generateOrderNo(),
                userId,
                totalAmount,
                OrderStatus.PENDING,
                now,
                now
        );
    }

    /**
     * 从持久化数据重建订单。仅供 infrastructure 层的 Converter 使用。
     */
    public static Order restore(Long id,
                                String orderNo,
                                Long userId,
                                BigDecimal totalAmount,
                                OrderStatus status,
                                LocalDateTime createdAt,
                                LocalDateTime updatedAt) {
        return new Order(id, orderNo, userId, totalAmount, status, createdAt, updatedAt);
    }

    /**
     * 持久化生成主键后由仓储调用一次：回填 id 并补登创建事件。
     */
    public void markCreated(Long generatedId) {
        if (this.id != null) {
            throw new IllegalStateException("订单 ID 已存在，不可重复初始化");
        }
        this.id = generatedId;
        recordEvent(new OrderCreatedEvent(
                this.id,
                this.orderNo,
                this.userId,
                this.totalAmount,
                this.status.name()
        ));
    }

    private static String generateOrderNo() {
        return "ORD" + System.currentTimeMillis();
    }

    public void cancel() {
        if (this.status != OrderStatus.PENDING) {
            throw new IllegalStateException("只有待支付订单可以取消");
        }
        this.status = OrderStatus.CANCELLED;
        this.updatedAt = LocalDateTime.now();

        recordEvent(new OrderCancelledEvent(
                this.id,
                this.orderNo,
                this.userId,
                this.totalAmount,
                this.status.name()
        ));
    }

    public void pay() {
        if (this.status != OrderStatus.PENDING) {
            throw new IllegalStateException("只有待支付订单可以支付");
        }
        this.status = OrderStatus.PAID;
        this.updatedAt = LocalDateTime.now();

        recordEvent(new OrderPaidEvent(
                this.id,
                this.orderNo,
                this.userId,
                this.totalAmount,
                this.status.name()
        ));
    }

    public void complete() {
        if (this.status != OrderStatus.PAID) {
            throw new IllegalStateException("只有已支付订单可以完成");
        }
        this.status = OrderStatus.COMPLETED;
        this.updatedAt = LocalDateTime.now();

        recordEvent(new OrderCompletedEvent(
                this.id,
                this.orderNo,
                this.userId,
                this.totalAmount,
                this.status.name()
        ));
    }

    private void recordEvent(DomainEvent event) {
        this.domainEvents.add(event);
    }

    public List<DomainEvent> getDomainEvents() {
        return Collections.unmodifiableList(new ArrayList<>(this.domainEvents));
    }

    public void clearDomainEvents() {
        this.domainEvents.clear();
    }
}

File: practice-projects/gin-ddd/internal/domain/model/order/order.go
// Package order 定义订单聚合根。
//
// 聚合根仅承载领域行为与不变量,不耦合任何持久化或框架细节。
// 持久化映射由 infrastructure 层的 OrderDO + Converter 完成。
package order

import (
	"fmt"
	"time"

	"gin-ddd/internal/domain/errors"
	"gin-ddd/internal/domain/event"
)

// Order 订单聚合根。字段保留公开(Go 风格),通过行为方法控制状态迁移,
// 外部代码不应直接赋值 Status / UpdatedAt 等字段。
type Order struct {
	OrderID     int64
	OrderNo     string
	UserID      int64
	TotalAmount float64
	Status      OrderStatus
	CreatedAt   time.Time
	UpdatedAt   time.Time

	// events 收集行为方法产生的领域事件,由应用层 PullEvents 后统一发布。
	// 字段私有,外部不可直接读写。
	events []event.DomainEvent
}

// OrderStatus 订单状态。
type OrderStatus string

const (
	OrderStatusPending   OrderStatus = "PENDING"   // 待支付
	OrderStatusPaid      OrderStatus = "PAID"      // 已支付
	OrderStatusShipped   OrderStatus = "SHIPPED"   // 已发货
	OrderStatusDelivered OrderStatus = "DELIVERED" // 已送达
	OrderStatusCancelled OrderStatus = "CANCELLED" // 已取消
	OrderStatusRefunded  OrderStatus = "REFUNDED"  // 已退款
)

// New 创建新订单。校验入参合法性,内部生成订单号,状态置为 PENDING。
// 不发布 OrderCreatedEvent——id 尚未生成,事件由 MarkCreated 在持久化后补登。
func New(userID int64, totalAmount float64) (*Order, error) {
	if userID <= 0 {
		return nil, errors.NewInvalidArgument("用户ID无效")
	}
	if totalAmount <= 0 {
		return nil, errors.NewInvalidArgument("订单金额无效")
	}

	now := time.Now()
	return &Order{
		OrderNo:     generateOrderNo(),
		UserID:      userID,
		TotalAmount: totalAmount,
		Status:      OrderStatusPending,
		CreatedAt:   now,
		UpdatedAt:   now,
	}, nil
}

// Restore 从持久化数据重建订单,仅供 infrastructure 层的 Converter 调用。
// 重建过程不产生领域事件。
func Restore(id int64, orderNo string, userID int64, totalAmount float64,
	status OrderStatus, createdAt, updatedAt time.Time) *Order {
	return &Order{
		OrderID:     id,
		OrderNo:     orderNo,
		UserID:      userID,
		TotalAmount: totalAmount,
		Status:      status,
		CreatedAt:   createdAt,
		UpdatedAt:   updatedAt,
	}
}

// MarkCreated 在仓储 save 完成、数据库回填主键后由仓储调用一次。
// 合并"赋 ID + 记录 OrderCreatedEvent",对齐 DDD"事件由聚合根记录"的约定。
func (o *Order) MarkCreated(id int64) error {
	if o.OrderID != 0 {
		return errors.NewInvalidState("订单 ID 已存在,不可重复初始化")
	}
	o.OrderID = id
	o.recordEvent(event.NewOrderCreatedEvent(o.OrderID, o.OrderNo, o.UserID, o.TotalAmount, string(o.Status)))
	return nil
}

// Pay 支付订单。
func (o *Order) Pay() error {
	if o.Status != OrderStatusPending {
		return errors.NewInvalidState("只有待支付订单可以支付")
	}
	o.Status = OrderStatusPaid
	o.UpdatedAt = time.Now()
	o.recordEvent(event.NewOrderPaidEvent(o.OrderID, o.OrderNo, o.UserID, o.TotalAmount, string(o.Status)))
	return nil
}

// Ship 发货。
func (o *Order) Ship() error {
	if o.Status != OrderStatusPaid {
		return errors.NewInvalidState("只有已支付订单可以发货")
	}
	o.Status = OrderStatusShipped
	o.UpdatedAt = time.Now()
	o.recordEvent(event.NewOrderShippedEvent(o.OrderID, o.OrderNo, o.UserID, o.TotalAmount, string(o.Status)))
	return nil
}

// Deliver 确认送达。
func (o *Order) Deliver() error {
	if o.Status != OrderStatusShipped {
		return errors.NewInvalidState("只有已发货订单可以确认送达")
	}
	o.Status = OrderStatusDelivered
	o.UpdatedAt = time.Now()
	o.recordEvent(event.NewOrderDeliveredEvent(o.OrderID, o.OrderNo, o.UserID, o.TotalAmount, string(o.Status)))
	return nil
}

// Cancel 取消订单。
func (o *Order) Cancel() error {
	if o.Status != OrderStatusPending {
		return errors.NewInvalidState("只有待支付订单可以取消")
	}
	o.Status = OrderStatusCancelled
	o.UpdatedAt = time.Now()
	o.recordEvent(event.NewOrderCancelledEvent(o.OrderID, o.OrderNo, o.UserID, o.TotalAmount, string(o.Status)))
	return nil
}

// Refund 退款。
func (o *Order) Refund() error {
	if o.Status != OrderStatusPaid && o.Status != OrderStatusShipped {
		return errors.NewInvalidState("只有已支付或已发货的订单可以退款")
	}
	o.Status = OrderStatusRefunded
	o.UpdatedAt = time.Now()
	o.recordEvent(event.NewOrderRefundedEvent(o.OrderID, o.OrderNo, o.UserID, o.TotalAmount, string(o.Status)))
	return nil
}

// IsPending 判断是否待支付。
func (o *Order) IsPending() bool { return o.Status == OrderStatusPending }

// IsPaid 判断是否已支付。
func (o *Order) IsPaid() bool { return o.Status == OrderStatusPaid }

// CanCancel 判断是否可取消。
func (o *Order) CanCancel() bool { return o.Status == OrderStatusPending }

// PullEvents 返回本次行为累积的领域事件,并清空内部列表。
// 应用层在调用 repository.Update 后通过此方法拿事件交给 EventPublisher。
func (o *Order) PullEvents() []event.DomainEvent {
	events := o.events
	o.events = nil
	return events
}

func (o *Order) recordEvent(e event.DomainEvent) {
	o.events = append(o.events, e)
}

func generateOrderNo() string {
	return fmt.Sprintf("ORD%d", time.Now().UnixNano())
}
